test/Microsoft.ML.TestFramework/BaseTestBaseline.cs

```diff
@@ -111,22 +111,41 @@ protected override void Initialize()
             ML.AddStandardComponents();
         }
 
-        private string GetConfigurationDir()
+        private IEnumerable<string> GetConfigurationDirs()
         {
+            // Sometimes different configuration will have combination.
+            // for example: one test can run on windows, have different result both
+            // on x64 vs x86 and dotnet core 3.1 vs others, so we have 4 combination:
+            // x64-netcore3.1, x86-netcore3.1, x64-rest, x86-rest. In some cases x64 vs x86
+            // have different results, in some cases netcore 3.1 vs rest have different results,
+            // the most complicate situation is these 4 combination have different results.
+            // So use list of string to return different configurations and test will try to search
+            // through this list and use the one first found, make sure we don't have baseline file
+            // at different configuration folders.
+            var configurationDirs = new List<string>();
+
             if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
-                return "osx-x64";
+                configurationDirs.Add("osx-x64");
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
-                return "linux-x64";
+                configurationDirs.Add("linux-x64");
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                 if (Environment.Is64BitProcess)
-                    return "win-x64";
+                    configurationDirs.Add("win-x64");
                 else
-                    return "win-x86";
+                    configurationDirs.Add("win-x86");
 
-            // We can add other configuration specific strings here for 
-            // tests that need configuration specific baselines
+            // Use netcore 3.1 result file if necessary.
+            // The small difference comes from CPUMath using different instruction set:
+            // 1. net framework and net core 2.1 uses CpuMathUtils.netstandard that uses SSE instruction set;
+            // 2. net core 3.1 uses CpuMathUtils.netcoreapp that uses AVX, SSE or direct floating point calculation
+            // depending on hardward avaibility.
+            // AVX and SSE generates slightly different result due to nature of floating point math.
+            // So Ideally we should adding AVX support at CPUMath native library, 
+            // use below issue to track: https://github.com/dotnet/machinelearning/issues/5044
+            if (AppDomain.CurrentDomain.GetData("FX_PRODUCT_VERSION") != null)
+                configurationDirs.Add("NetCore3.1Result");
 
```

Comment:
This is a good change overall to move to multiple configuration dirs. 

What happens if we have stored different results for win-x64-netcoreapp21, win-x64-netcoreapp31, win-x86-netcoreapp31, winx-x86-netcoreapp21?

Even if we store netcoreapp21 results by default into win-x64 and win-x86, we would still 2 separate directories for win-x64-netcoreapp31 and win winx86-netcoreapp31.

Does that sound correct? Or am I missing something?

Also can you please change the directory name to be just reflect the configuration (e.g. netcoreapp31 as opposed to NetCore3.1Result)?

Reply:
Yes, I can change the folder name to netcoreapp31.

What happens if we have stored different results for win-x64-netcoreapp21, win-x64-netcoreapp31, win-x86-netcoreapp31, winx-x86-netcoreapp21?

 - in that case, we have to return win-x64-netcoreapp21, win-x64-netcoreapp31, win-x86-netcoreapp31, winx-x86-netcoreapp21 as configuration directory and make sure we don't have any "expected" result file at folder win-64 or netcoreapp31. And during get file stage we will check one by one if file exists. Let me make the logic complete and send out new iteration.


---
In reply to: [413162621](https://github.com/dotnet/machinelearning/pull/5047#discussion_r413162621) [](ancestors = 413162621)